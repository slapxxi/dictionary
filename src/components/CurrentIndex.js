// @flow
import _ from 'lodash';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { changeIndex } from '../store/actions';

type Props = {
  index: number,
  total: number,
  style?: Object,
  onChange?: (number) => void,
};

type State = { active: boolean };

class CurrentIndex extends Component<Props, State> {
  state = { active: false };

  handleClick = () => {
    this.setState({ active: true });
  };

  handleBlur = ({ currentTarget }: Event) => {
    const { index, onChange } = this.props;
    if (onChange) {
      const nextIndex = this.parseIndex(currentTarget.value);
      if (nextIndex && nextIndex !== index + 1) {
        onChange(nextIndex);
      }
    }
    this.setState({ active: false });
  };

  handleChange = (event: Event) => {
    if (event.key === 'Enter') {
      this.handleBlur(event);
    }
  };

  parseIndex = (value: string) => {
    const parsed = Number.parseInt(value, 10);
    if (Number.isNaN(parsed)) {
      return undefined;
    }
    return this.clamp(parsed);
  };

  clamp = (value: number) => {
    const { total } = this.props;
    return _.clamp(value, 1, total);
  };

  render() {
    const { style, index, total } = this.props;
    if (this.state.active) {
      return this.renderInput();
    }
    return (
      <nav style={style} className="current-index">
        <button onClick={this.handleClick}>
          {index + 1}/{total}
        </button>
      </nav>
    );
  }

  renderInput() {
    const { style } = this.props;
    return (
      <input
        autoFocus
        style={style}
        type="text"
        onKeyPress={this.handleChange}
        onBlur={this.handleBlur}
        className="current-index_input"
      />
    );
  }
}

const enhance = connect(
  ({ dictionary }) => ({
    index: dictionary.index,
    total: dictionary.entries.length,
  }),
  { onChange: (index) => changeIndex(index - 1) },
);

export { CurrentIndex };

export default enhance(CurrentIndex);
