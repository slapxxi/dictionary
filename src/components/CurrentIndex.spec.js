import React from 'react';
import { shallow } from 'enzyme';
import { CurrentIndex } from './CurrentIndex';

it('renders', () => {
  const component = shallow(<CurrentIndex index={0} total={20} />);
  expect(component).toMatchSnapshot();
});

it('accepts style', () => {
  const component = shallow(
    <CurrentIndex index={10} total={20} style={{ flex: 1 }} />,
  );
  expect(component).toMatchSnapshot();
});

describe('given input is active', () => {
  let component;
  let input;
  let onChange;

  beforeEach(() => {
    onChange = jest.fn();
    component = shallow(
      <CurrentIndex index={3} total={10} onChange={onChange} />,
    );
    component.find('.current-index > button').simulate('click');
    input = component.find('.current-index_input');
  });

  describe('when loses focus', () => {
    it('invokes onChange callback', () => {
      input.simulate('blur', { currentTarget: { value: '10' } });
      expect(onChange).toBeCalledWith(10);
    });

    it('does not invoke callback when changed to text', () => {
      input.simulate('blur', { currentTarget: { value: 'test' } });
      expect(onChange).not.toBeCalled();
    });

    it('does not invoke callback when changed to same value', () => {
      input.simulate('blur', { currentTarget: { value: '4' } });
      expect(onChange).not.toBeCalled();
    });

    it('passes value clamped to upper limit', () => {
      input.simulate('blur', { currentTarget: { value: '20' } });
      expect(onChange).toBeCalledWith(10);
    });

    it('passes value clamped to lower limit', () => {
      input.simulate('blur', { currentTarget: { value: '-10' } });
      expect(onChange).toBeCalledWith(1);
    });
  });
});
