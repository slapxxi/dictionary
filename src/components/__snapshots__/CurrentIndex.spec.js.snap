// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts style 1`] = `
<nav
  className="current-index"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <button
    onClick={[Function]}
  >
    11
    /
    20
  </button>
</nav>
`;

exports[`renders 1`] = `
<nav
  className="current-index"
>
  <button
    onClick={[Function]}
  >
    1
    /
    20
  </button>
</nav>
`;
